src/main/java/ch/njol/skript/variables/FlatFileStorage.java

```diff
@@ -200,184 +279,154 @@ public void run() {
-	}
-	
+
 	@Override
-	protected boolean save(final String name, final @Nullable String type, final @Nullable byte[] value) {
-		synchronized (connectionLock) {
-			synchronized (changesWriter) {
-				if (!loaded && type == null)
-					return true; // deleting variables is not really required for this kind of storage, as it will be completely rewritten every once in a while, and at least once when the server stops.
-				PrintWriter cw;
-				while ((cw = changesWriter.get()) == null) {
-					try {
-						changesWriter.wait();
-					} catch (final InterruptedException e) {
-						Thread.currentThread().interrupt();
-					}
-				}
-				writeCSV(cw, name, type, value == null ? "" : encode(value));
-				cw.flush();
-				changes.incrementAndGet();
-			}
-		}
-		return true;
+	protected File getFile(String fileName) {
+		return new File(fileName);
 	}
-	
-	/**
-	 * Use with find()
-	 */
-	@SuppressWarnings("null")
-	private final static Pattern containsWhitespace = Pattern.compile("\\s");
-	
-	private static void writeCSV(final PrintWriter pw, final String... values) {
-		assert values.length == 3; // name, type, value
-		for (int i = 0; i < values.length; i++) {
-			if (i != 0)
-				pw.print(", ");
-			String v = values[i];
-			if (v != null && (v.contains(",") || v.contains("\"") || v.contains("#") || containsWhitespace.matcher(v).find()))
-				v = '"' + v.replace("\"", "\"\"") + '"';
-			pw.print(v);
-		}
-		pw.println();
-	}
-	
+
 	@Override
 	protected final void disconnect() {
 		synchronized (connectionLock) {
 			clearChangesQueue();
 			synchronized (changesWriter) {
-				final PrintWriter cw = changesWriter.get();
-				if (cw != null) {
-					cw.close();
+				PrintWriter printWriter = changesWriter.get();
+
+				if (printWriter != null) {
+					printWriter.close();
 					changesWriter.set(null);
 				}
 			}
 		}
 	}
-	
+
 	@Override
 	protected final boolean connect() {
 		synchronized (connectionLock) {
 			synchronized (changesWriter) {
+				assert file != null; // file should be non-null after load
+
 				if (changesWriter.get() != null)
 					return true;
-				try (FileOutputStream fos = new FileOutputStream(file, true)){
-					changesWriter.set(new PrintWriter(new OutputStreamWriter(fos, UTF_8)));
+
+				// Open the file stream, and create the PrintWriter with it
+				try (FileOutputStream fos = new FileOutputStream(file, true)) {
+					changesWriter.set(new PrintWriter(new OutputStreamWriter(fos, FILE_CHARSET)));
 					loaded = true;
 					return true;
 				} catch (IOException e) { // close() might throw ANY IOException
+					//noinspection ThrowableNotThrown
 					Skript.exception(e);
 					return false;
 				}
 			}
 		}
 	}
-	
+
 	@Override
 	public void close() {
 		clearChangesQueue();
 		super.close();
 		saveVariables(true); // also closes the writer
 	}
-	
+
+	@Override
+	protected boolean save(String name, @Nullable String type, @Nullable byte[] value) {
+		synchronized (connectionLock) {
+			synchronized (changesWriter) {
```

Comment:
I'm concerned we might deadlock here if something asks for connection lock in order to wake the writer, but since this is old code it doesn't seem to have been a problem.

Reply:
Does Java deadlock when trying to acquire a lock which has already been acquired somewhere up the stack? I thought it had sth for that

Reply:
I can't remember if the outer lock will become available for the purposes of notifying when the inner one is waiting, I'll have to check.